components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java

```diff
@@ -0,0 +1,481 @@
+     * Verify the JWT token signature.
+     *
+     * @param splitToken The JWT token which is split into [header, payload, signature]
+     * @return whether the signature is verified or or not
+     * @throws APISecurityException in case of signature verification failure
+     */
+    private boolean verifyTokenSignature(String[] splitToken) throws APISecurityException {
+        // Retrieve signature algorithm from token header
+        String signatureAlgorithm = getSignatureAlgorithm(splitToken);
+
+        Certificate publicCert;
+        //Read the client-truststore.jks into a KeyStore
+        try {
+            KeyStore trustStore = ServiceReferenceHolder.getInstance().getTrustStore();
+            // Read public certificate from trust store
+            publicCert = trustStore.getCertificate(APIConstants.GATEWAY_PUBLIC_CERTIFICATE_ALIAS);
+        } catch (KeyStoreException e) {
+            throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                    "Error in retrieving public certificate from the trust store with alias : "
+                            + APIConstants.GATEWAY_PUBLIC_CERTIFICATE_ALIAS, e);
+        }
+
+        if (publicCert != null) {
+            // Retrieve public key from the certificate
+            PublicKey publicKey = publicCert.getPublicKey();
+
+            try {
+                // Verify token signature
+                Signature signatureInstance = Signature.getInstance(signatureAlgorithm);
+                signatureInstance.initVerify(publicKey);
+                String assertion = splitToken[0] + "." + splitToken[1];
+                signatureInstance.update(assertion.getBytes());
+                byte[] decodedSignature = Base64.getUrlDecoder().decode(splitToken[2]);
+                return signatureInstance.verify(decodedSignature);
+            } catch (NoSuchAlgorithmException | InvalidKeyException | SignatureException | IllegalArgumentException e) {
+                log.error("Signature verification failed.", e);
+                throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
+                        "Invalid JWT token", e);
+            }
+        } else {
+            throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                    "Couldn't find a public certificate to verify signature");
+        }
+    }
+
+    /**
+     * Retrieve the signature algorithm specified in the token header.
+     *
+     * @param splitToken The JWT token which is split into [header, payload, signature]
+     * @return whether the signature algorithm
+     * @throws APISecurityException in case of signature algorithm extraction failure
+     */
+    private String getSignatureAlgorithm(String[] splitToken) throws APISecurityException {
+        String signatureAlgorithm;
+        JSONObject header;
+        try {
+            header = new JSONObject(new String(Base64.getUrlDecoder().decode(splitToken[0])));
```

Comment:
seems we do url decode? Can there be any other decoding methods?

Reply:
In the Key manager, we do URL encode. That's why I used URL decode here.